x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java

```diff
@@ -605,10 +613,10 @@ public void closeJob(JobTask jobTask, boolean restart, String reason) {
             if (communicator == null) {
                 logger.debug("Job [{}] is being closed before its process is started", jobId);
                 jobTask.markAsCompleted();
-                return;
```

Comment:
Returning here without removing the entry from `processByAllocation` is the cause of the eventual filling up to capacity of `processByAllocation`.  It's nasty because `jobTask.markAsCompleted()` causes the externally visible effects of closing the job to all look as expected.  It's only the internal data structure that's left with a spurious entry.